x-pack/legacy/plugins/ml/public/data_frame/pages/transform_management/components/transform_list/transform_list.tsx

```diff
@@ -76,6 +90,75 @@ export const DataFrameTransformList: SFC = () => {
   // Call useRefreshInterval() after the subscription above is set up.
   useRefreshInterval(setBlockRefresh);
 
+  const onQueryChange = ({ query }: { query: Query }) => {
+    setIsLoading(true);
+    const clauses = query.ast.clauses;
+
+    if (clauses.length === 0) {
+      setFilteredTransforms(transforms);
+      setIsLoading(false);
+      return;
+    }
+    // keep count of the number of matches we make as we're looping over the clauses
+    // we only want to return transforms which match all clauses, i.e. each search term is ANDed
+    // { transform-one:  { transform: { id: transform-one, config: {}, state: {}, ... }, count: 0 }, transform-two: {...} }
+    const matches: { [key: string]: any } = transforms.reduce((p: { [key: string]: any }, c) => {
+      p[c.id] = {
+        transform: c,
+        count: 0,
+      };
+      return p;
+    }, {});
+
+    clauses.forEach(c => {
+      // the search term could be negated with a minus, e.g. -bananas
+      const bool = c.match === 'must';
+      let ts = [];
+
+      if (c.type === 'term') {
+        // filter term based clauses, e.g. bananas
+        // match on id and description
+        // if the term has been negated, AND the matches
+        if (bool === true) {
+          ts = transforms.filter(
+            transform =>
+              stringMatch(transform.id, c.value) === bool ||
+              stringMatch(transform.config.description, c.value) === bool
+          );
+        } else {
+          ts = transforms.filter(
+            transform =>
+              stringMatch(transform.id, c.value) === bool &&
+              stringMatch(transform.config.description, c.value) === bool
+          );
+        }
+      } else {
+        // filter other clauses, i.e. the mode and status filters
+        if (Array.isArray(c.value)) {
+          // the status value is an array of string(s) e.g. ['failed', 'stopped']
+          ts = transforms.filter(transform => c.value.includes(transform.state.task_state));
+        } else {
+          ts = transforms.filter(transform => transform.config.mode === c.value);
+        }
+      }
+
+      ts.forEach(t => matches[t.id].count++);
+    });
+
+    // loop through the matches and return only those jobs which have match all the clauses
+    const filtered = [];
+    for (const key in matches) {
```

Comment:
could be written as
```
const filtered = Object.values(matches)
  .filter(m => m.count >= clauses.length)
  .map(m => m.transform);
```

Reply:
Excellent suggestion 👌 Here's the change: 7f77e964edc9b29b479a3bf9b9b06bba1a05122a